torch/distributed/traceable_collectives.py

```diff
@@ -0,0 +1,202 @@
+def _all_reduce(self, reduceOp, tag, ranks, stride):
+    reduceOp = reduceOp.lower()
+    assert reduceOp == "sum", "Unable to convert str to ReduceOp, so only default sum works"
+    assert tag == "", "No support for non-empty comms tag"
+    assert len(ranks) % stride == 0, f"Ranks length ({len(ranks)}) must be divisible by stride ({stride})"
+
+    my_rank = dist.get_rank()
+    rank_set = None
+
+    for i in range(0, len(ranks), stride):
+        rank_set = ranks[i : i + stride]
+        if my_rank in rank_set:
+            my_ranks = rank_set
+
+    assert my_ranks is not None, "Called all_reduce with a set of ranks that doesn't include the current node"
+    assert my_rank in my_ranks, "Called all_reduce with a set of ranks that doesn't include the current node"
+
+    my_ranks.sort()
+
+    group = c10d._find_pg_by_ranks_and_tag(tag, my_ranks)
+    assert group is not None
+
+    inplace_tensor = self.clone()
+    work = dist.all_reduce(inplace_tensor, op=dist.ReduceOp.SUM, group=group, async_op=True)
+
+    global tensor_to_work
+    tensor_to_work.add(inplace_tensor, work)
+    return inplace_tensor
+
+c10_lib_cpu = torch.library.Library("aten", "IMPL", "CPU")
+c10_lib_cuda = torch.library.Library("aten", "IMPL", "CUDA")
+
+c10_lib_cpu.impl("all_reduce", _all_reduce)
+c10_lib_cuda.impl("all_reduce", _all_reduce)
+
+RANK_TYPES = Union[List[int], List[List[int]], dist.ProcessGroup, dt.DeviceMesh, Tuple[dt.DeviceMesh, int]]
+
+def _expand_group(group: RANK_TYPES, tag: str = "") -> Tuple[str, List[int], int]:
+    rankset: List[int]
+    if isinstance(group, list):
+        if isinstance(group[0], list):
+            nested_list = cast(List[List[int]], group)
+            rankset = []
+            for rs in nested_list:
+                rankset.extend(rs)
+                stride = len(rs)
+        else:
+            rankset = cast(List[int], group)
+            stride = len(rankset)
+    elif isinstance(group, dist.ProcessGroup):
+        rankset = dist.get_process_group_ranks(group)
+        stride = len(rankset)
+        tag = tag or c10d._get_group_tag(group)
+    elif isinstance(group, dt.DeviceMesh):
+        rankset = group.mesh.flatten().tolist()
+        stride = len(rankset)
+        tag = tag or c10d._get_group_tag(group.get_dim_groups()[0])
+    elif isinstance(group, tuple):
+        if len(group) == 2 and isinstance(group[0], dt.DeviceMesh) and isinstance(group[1], int):
+            dmesh = group[0]
+            dim = group[1]
+            stride = dmesh.mesh.size(dim)
+            rankset = dmesh.mesh.swapdims(-1, dim).reshape(-1, stride).flatten().tolist()
+            tag or c10d._get_group_tag(dmesh.get_dim_groups()[dim])
+        else:
+            raise ValueError("Invalid type for group")
+    else:
+        raise ValueError("Invalid type for group")
+
+    return (tag, rankset, stride)
+
+def all_reduce(self: torch.Tensor, reduceOp: str, group: RANK_TYPES, tag: str = ""):
```

Comment:
it looks like you added another api layer compared to my initial design.  can you walk through the reasoning?

what are the flows you are thinking of?

1) dist.all_reduce -> tr.all_reduce -> aten.all_reduce
2) DTensor.traceable -> tr.all_reduce -> aten.all_reduce
3) user code -> tr.all_reduce ->  aten.all_reduce

any reason to not just embed the 'AsyncCollectiveTensor wrapping' inside aten.all_reduce impl?

actually maybe i get it now: this is how you make dispatch tracing work- if everyone calls tr.all_reduce, we then get to re-wrap the faketensor output of aten.Allreduce with a subclass. if you do it inside  aten.allreduce instead, i think that will make it so tracing does not capture wait op.  This explains part of my confusion from the design doc.

So to follow up on that, do you actually think we should support flow (1) or skip it?

Reply:
We are ignoring (1) for now until we find a scenario where it's worth considering it.

I think the way to support one is to make it an opt-in feature for users that, for some reason, can't just rewrite the code in question.